### DISCUSSION TITLE
Rubrik Capstone Bike Sharing API

### DISCUSSION THREAD

**Discussion by Putrareza** [teamalgoritma/community#865] [Q&A] [answered]
Maaf mau bertnya gimana cara nyelesain 3 rubrik terakhir Capstone Bike Sharing API?, soalnya aku terkendala dan bingung mau mulai dari mana

**Comment by IrfanChairurrachman** [comment]
Dear @Putrareza ,

Untuk mengerjakan 3 rubrik terakhir, Anda dapat membuat endpoint yang dapat melakukan analisa data, bentuk analisanya bebas, diserahkan kepada Anda selama endpoint yang diminta berhasil me-_return_ hasil analisa.

![image](https://github.com/teamalgoritma/community/assets/56690035/61edb8a3-5f69-4813-a27e-09a21d17ecd9)

Sebagai contoh, pada endpoint statis Anda membuat endpoint untuk return nilai average_duration sepeda pada bikeshare. Maka tahapannya:

Membuat endpoint statis untuk trips/average_duration di app.py
Membuat fungsi query untuk mendapatkan hasil average_duration di app.py
Menjalankan app.py
Memastikan endpoint yang dibuat berhasil dengan melakukan testing di jupyter notebook pada section "TEST YOUR ENDPOINT" (seperti gambar di bawah). Jangan lupa untuk mengganti url endpoint sesuai dengan endpoint yang telah dibuat.

![image](https://github.com/teamalgoritma/community/assets/56690035/1f2f366e-47bc-4c44-ad9e-ac45f9ad099b)

Patut dicatat bahwa average_duration ini hanyalah contoh, Anda bisa membuat analisis yang lain seperti mendapatkan jumlah keseluruhan sepeda pada semua stasiun atau analisis lainnya.

Sekian dari kami,
Semoga membantu

**Reply by Putrareza** [reply]
Apakah Saya bisa langsung membuat endpoit baru tanpa harus mengetes atau membuat endpoint avareges_duration?

**Reply by IrfanChairurrachman** [reply]
dear @Putrareza ,

Anda dapat membuat langsung endpoint baru. Namun sebaiknya di test di notebook untuk memastikan endpoint Anda berjalan dengan baik.

**Reply by Putrareza** [reply]
Berarti buat dari FLASK ya?

**Reply by Putrareza** [reply]
aku bingungnya au start dari mana aku nemu di vidio hanya yang dinamis aku gak nemu yang mana yang statis sedangkan penjelasannya kan harus buat dari endpoint statis jadi aku bingung endpoint statis itu kayak giimana

**Reply by IrfanChairurrachman** [reply]
Jika 3 rubrik pertama telah diselesaikan, bisa menggunakan konsep yang sama untuk 3 rubrik terakhir. Jika dirasa dapat membuat dynamic endpoint terlebih dahulu kemudian baru statis endpoint tidak ada masalah kok Pak selama semua rubrik bisa diselesaikan.

Untuk video pembuatan statis dan dinamis dapat menerapkan konsep dari video nomor 7, adapun penjelasan untuk ke-3 rubrik terakhir terdapat pada video terakhir, yaitu video ke-9.

Semoga membantu Pak

**Reply by Putrareza** [reply]
Mas ngomong ngomong yang vidio 7 itu hanya menjelaskan endpoint dynamic aja gak ada dia jelasin yang statis

**Reply by Putrareza** [reply]
![Screenshot 2024-04-29 185506](https://github.com/teamalgoritma/community/assets/60474540/5bb3d040-ab44-46bb-9cba-a4085609c2ab)

**Reply by Putrareza** [reply]
berarti kita ttp menggunakan fungsi endpoint statis yang fungsi "/" dan "/trips"
sisanya untuk yang dynamic kita buat analisa sendiri misal kita mau membuat endpoint berdasarkan bike_id? benerkah analisa saya ini mas?

**Reply by Putrareza** [reply]
apakah saya boleh menggunakan endpoint yang sama pada contoh yang endpoint /trips/averages_duration  ?

**Reply by Putrareza** [reply]
![image](https://github.com/teamalgoritma/community/assets/60474540/1f1b1b2b-57ad-4e8e-9afe-0160bf6bb988)
pas mau coba test running yang dynamic endpoint selalu responnya 500 knp ya?

**Reply by IrfanChairurrachman** [reply]
Dear @Putrareza ,

Jika diperhatikan dari pesan errornya terdapat "missing 1 required positional argument" atau terdapat argument/nilai dinamis yg tidak diterima oleh endpoint. Endpoint yang bersifat dinamis contohnya seperti endpoint untuk mendapatkan stasiun tertentu. Seperti gambar di bawah.

![image](https://github.com/teamalgoritma/community/assets/56690035/0d44b857-bb51-4d1a-9d6f-5aba7a72a350)

Dimana url menerima inputan <station_id>, jika dicontohkan kita ingin mendapatkan informasi mengenai stasiun 3464, sehingga ketika di test akan seperti berikut

![image](https://github.com/teamalgoritma/community/assets/56690035/c49eaee9-aa4f-4503-8305-4696b416e237)

Maka dari itu, silahkan untuk membuat endpoint dinamis untuk menganalisis data, seperti mendapatkan jumlah sepeda per stasiunnya.

Sekian, semoga membantu

**Reply by Putrareza** [reply]
Alhamdulliah kalau masalah endpoint dinamis udah aku dapat jalan keluarnya
Tpi yang jadi maslah itu skrng yang create post endpoints yang soal terakhir
boleh saya kasih contoh gak cara pengerjaan selain yang di vidio sama jupiter notebook
soalnya udah 10x saya putar ualng ulang vidionya saya gak dapat dapat jalan cara pengerjaannya

**Reply by IrfanChairurrachman** [reply]
Alhamdulillah,

Untuk rubrik POST yang terakhir bisa dikembangkan dari endpoint JSON yg telah dibuat
![image](https://github.com/teamalgoritma/community/assets/56690035/466fdb6b-36e1-462e-b15d-6821f3c94540)

Endpoint JSON menerima inputan berupa `name`, `age`, dan `address` sehingga jika dipanggil user harus provide info tersebut
![image](https://github.com/teamalgoritma/community/assets/56690035/3ff2de03-e57a-4f76-8350-029402615827)


Nah, maka dari itu, Anda dapat membuat endpoint POST dimana menerima inputan, semisal inputan tahun untuk mencari jumlah sepeda yang dirental dalam satu tahun.

Semoga membantu

**Reply by Putrareza** [reply]
OOooo i see oke siap , paham sudah aku berarti dia hanya di test lewat Phyton sama postman gak di app.py benerkan mas irfan?

**Reply by IrfanChairurrachman** [reply]
iya betul Pak, tinggal diimplementasikan di app.py. Kemudian untuk testing bisa di postman atau notebooknya juga pada bagian "TEST YOUR ENDPOINT" (paling bawah di template notebook)

**Reply by Putrareza** [reply]
![image](https://github.com/teamalgoritma/community/assets/60474540/1d9e03ac-d37b-4d34-8e0c-d54f4dbd8c13)
tpi knp ya punya saya error pas mau coba show data frame, apakah coding saya yang salah atau gimana?

**Reply by IrfanChairurrachman** [reply, marked as answer]
Dear @Putrareza ,

untuk POST Endpoint hasil yang diharapkan berupa dataframe yang berasal dari hasil query di database (mirip seperti dynamic endpoint).

Jadi setelah mendapatkan inputan dari POST, inputannya menjadi parameter untuk dilakukan query. Jadi konsepnya mirip seperti dynamic endpoint hanya saja inputannya menggunakan POST alih-alih menggunakan endpoint dinamis.
